scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala

```diff
@@ -749,12 +750,88 @@ class PairSCollectionFunctions[K, V](val self: SCollection[(K, V)]) {
+   * buffered can be set. Once a batch is flushed to output, the timer is reset. The provided limit
+   * must be a positive duration or zero; a zero buffering duration effectively means no limit.
+   *
    * @param batchSize
+   * @param maxBufferingDuration
+   *
+   * @group per_key
+   */
+  def batchByKey(
+    batchSize: Long,
+    maxBufferingDuration: Duration = Duration.ZERO
+  ): SCollection[(K, Iterable[V])] = {
+    val groupIntoBatches = GroupIntoBatches
+      .ofSize[K, V](batchSize)
+      .withMaxBufferingDuration(maxBufferingDuration)
+    this.applyPerKey(groupIntoBatches)(kvIterableToTuple)
+  }
+
+  /**
+   * Batches inputs to a desired batch of byte size. Batches will contain only elements of a single
+   * key.
+   *
+   * The value coder is used to determine the byte size of each element.
+   *
+   * Elements are buffered until there are an estimated batchByteSize bytes buffered, at which point
+   * they are outputed to the output [[SCollection]].
+   *
+   * Windows are preserved (batches contain elements from the same window). Batches may contain
+   * elements from more than one bundle.
+   *
+   * A time limit (in processing time) on how long an incomplete batch of elements is allowed to be
+   * buffered can be set. Once a batch is flushed to output, the timer is reset. The provided limit
+   * must be a positive duration or zero; a zero buffering duration effectively means no limit.
+   *
+   * @param batchByteSize
+   * @param maxBufferingDuration
    *
    * @group per_key
    */
-  def batchByKey(batchSize: Long): SCollection[(K, Iterable[V])] =
-    this.applyPerKey(GroupIntoBatches.ofSize(batchSize))(kvIterableToTuple)
+  def batchByteSizedByKey(
+    batchByteSize: Long,
+    maxBufferingDuration: Duration = Duration.ZERO
+  ): SCollection[(K, Iterable[V])] = {
+    val groupIntoBatches = GroupIntoBatches
+      .ofByteSize[K, V](batchByteSize)
+      .withMaxBufferingDuration(maxBufferingDuration)
+    this.applyPerKey(groupIntoBatches)(kvIterableToTuple)
+  }
+
+  /**
+   * Batches inputs to a desired weight. Batches will contain only elements of a single key.
+   *
+   * The weight of each element is computer from the provided cost function.
+   *
+   * Elements are buffered until the weight is reached, at which point they are outputed to the
+   * output [[SCollection]].
+   *
+   * Windows are preserved (batches contain elements from the same window). Batches may contain
+   * elements from more than one bundle.
+   *
+   * A time limit (in processing time) on how long an incomplete batch of elements is allowed to be
+   * buffered can be set. Once a batch is flushed to output, the timer is reset. The provided limit
+   * must be a positive duration or zero; a zero buffering duration effectively means no limit.
+   *
+   * @param weight
+   * @param cost
+   * @param maxBufferingDuration
+   *
+   * @group per_key
+   */
+  def batchWeightedByKey(
```

Comment:
just semantically, this function name kind of seems like "per key" refers to the weighting, not the batching ... wdyt about something like `weightedBatchByKey`/`byteSizeBatchByKey`? I guess the latter sounds a bit clunky.

Reply:
+1 to keeping the prefix. Another alternative is keeping all `batchByKey` as the prefix so we'll get `batchByKeyWeighted` and `batchByKeyByteSized` to avoid the confusion Claire mentioned.